rust_src/src/eval.rs

```diff
@@ -1334,4 +1330,158 @@ pub fn condition_case(args: LispCons) -> LispObject {
     unsafe { internal_lisp_condition_case(var, bodyform, handlers) }
 }
 
+// Eval a sub-expression of the current expression (i.e. in the same lexical scope).
+#[no_mangle]
+pub extern "C" fn eval_sub(form: LispObject) -> LispObject {
+    if let Some(symbol) = form.as_symbol() {
+        let env = unsafe { globals.Vinternal_interpreter_environment };
+
+        // Look up its binding in the lexical environment.
+        // We do not pay attention to the declared_special flag here, since we
+        // already did that when let-binding the variable.
+        let lex_binding = env.map_or(Qnil, |e| assq(form, e));
+
+        if let Some((_, cdr)) = lex_binding.into() {
+            return cdr;
+        } else {
+            return symbol_value(symbol);
+        }
+    }
+
+    let (original_fun, mut original_args) = match form.as_cons() {
+        Some(cons) => cons.into(),
+        None => {
+            return form;
+        }
+    };
+
+    unsafe {
+        maybe_quit();
+
+        maybe_gc();
+
+        let mut current_thread = ThreadState::current_thread();
+        current_thread.m_lisp_eval_depth += 1;
+
+        if current_thread.m_lisp_eval_depth > globals.max_lisp_eval_depth {
+            if globals.max_lisp_eval_depth < 100 {
+                globals.max_lisp_eval_depth = 100;
+            }
+            if current_thread.m_lisp_eval_depth > globals.max_lisp_eval_depth {
+                error!("Lisp nesting exceeds `max-lisp-eval-depth'");
+            }
+        }
+
+        original_args.check_list();
+
+        // This also protects them from gc.
+        let count = record_in_backtrace(original_fun, &mut original_args, UNEVALLED as isize);
+
+        if globals.debug_on_next_call {
+            do_debug_on_call(Qt, count);
+        }
+
+        // At this point, only count, original_fun, and original_args
+        // have values that will be used below.
+        let (mut val, done) = eval_sub_1(original_fun, original_args, count);
+        if done {
+            return val;
+        }
+
+        check_cons_list();
+
+        current_thread.m_lisp_eval_depth -= 1;
+        if backtrace_debug_on_exit(current_thread.m_specpdl.offset(count)) {
+            val = call_debugger(list!(Qexit, val));
+        }
+
+        current_thread.m_specpdl_ptr = current_thread.m_specpdl_ptr.offset(-1);
+
+        val
+    }
+}
+
+// Actual work for evaluating a sub-expression happens here.
+fn eval_sub_1(
+    original_fun: LispObject,
+    original_args: LispObject,
+    count: libc::ptrdiff_t,
+) -> (LispObject, bool) {
+    let mut keep_going = true;
+    let mut val = Qnil;
+
+    while keep_going {
```

Comment:
The `while` can be changed to `loop` and `keep_going` can be dropped. The `keep_going = false` line can be replaced with `break` or `return (val, false)`.

Reply:
I tried that originally and was unhappy with it. If you have time, would you give it a go and either share a diff or directly edit this PR?

Reply:
```diff
modified   rust_src/src/eval.rs
@@ -1407,10 +1407,9 @@ fn eval_sub_1(
     original_args: LispObject,
     count: libc::ptrdiff_t,
 ) -> (LispObject, bool) {
-    let mut keep_going = true;
     let mut val = Qnil;
 
-    while keep_going {
+    loop {
         // Optimize for no indirection.
         let mut fun = original_fun;
         if let Some(symbol) = fun.as_symbol() {
@@ -1478,10 +1477,8 @@ fn eval_sub_1(
             }
         }
 
-        keep_going = false;
+        return (val, false);
     }
-
-    (val, false)
 }
 
 include!(concat!(env!("OUT_DIR"), "/eval_exports.rs"));
```

I think that should do it.